File: api/src/main/java/org/apache/iceberg/transforms/Dates.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg.transforms;

import com.google.errorprone.annotations.Immutable;
import java.time.temporal.ChronoUnit;
import org.apache.iceberg.expressions.BoundPredicate;
import org.apache.iceberg.expressions.BoundTransform;
import org.apache.iceberg.expressions.Expression;
import org.apache.iceberg.expressions.Expressions;
import org.apache.iceberg.expressions.UnBoundCreator;
import org.apache.iceberg.expressions.UnboundPredicate;
import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
import org.apache.iceberg.types.Type;
import org.apache.iceberg.types.Types;
import org.apache.iceberg.util.DateTimeUtil;
import org.apache.iceberg.util.SerializableFunction;

enum Dates implements Transform<Integer, Integer> {
  YEAR(ChronoUnit.YEARS, "year"),
  MONTH(ChronoUnit.MONTHS, "month"),
  DAY(ChronoUnit.DAYS, "day");

  @Immutable
  static class Apply implements SerializableFunction<Integer, Integer> {
    private final ChronoUnit granularity;

    Apply(ChronoUnit granularity) {
      this.granularity = granularity;
    }

    @Override
    public Integer apply(Integer days) {
      if (days == null) {
        return null;
      }

      switch (granularity) {
        case YEARS:
          return DateTimeUtil.daysToYears(days);
        case MONTHS:
          return DateTimeUtil.daysToMonths(days);
        case DAYS:
          return days;
        default:
          throw new UnsupportedOperationException("Unsupported time unit: " + granularity);
      }
    }
  }

  private final ChronoUnit granularity;
  private final String name;
  private final Apply apply;

  Dates(ChronoUnit granularity, String name) {
    this.granularity = granularity;
    this.name = name;
    this.apply = new Apply(granularity);
  }

  @Override
  public Integer apply(Integer days) {
    return apply.apply(days);
  }

  @Override
  public SerializableFunction<Integer, Integer> bind(Type type) {
    Preconditions.checkArgument(canTransform(type), "Cannot bind to unsupported type: %s", type);
    return apply;
  }

  @Override
  public boolean canTransform(Type type) {
    return type.typeId() == Type.TypeID.DATE;
  }

  @Override
  public Type getResultType(Type sourceType) {
    if (granularity == ChronoUnit.DAYS) {
      return Types.DateType.get();
    }
    return Types.IntegerType.get();
  }

  @Override
  public boolean preservesOrder() {
    return true;
  }

  @Override
  public boolean satisfiesOrderOf(Transform<?, ?> other) {
    if (this == other) {
      return true;
    }

    if (other instanceof Dates) {
      // test the granularity, in days. day(ts) => 1 day, months(ts) => 30 days, and day satisfies
      // the order of months
      Dates otherTransform = (Dates) other;
      return granularity.getDuration().toDays()
          <= otherTransform.granularity.getDuration().toDays();
    }

    return false;
  }

  @Override
  public UnboundPredicate<Integer> project(String fieldName, BoundPredicate<Integer> pred) {
    if (pred.term() instanceof BoundTransform) {
      return ProjectionUtil.projectTransformPredicate(this, fieldName, pred);
    }

    if (pred.isUnaryPredicate()) {
      return Expressions.predicate(pred.op(), fieldName);

    } else if (pred.isLiteralPredicate()) {
      UnboundPredicate<Integer> projected =
          ProjectionUtil.truncateInteger(fieldName, pred.asLiteralPredicate(), apply);
      if (this != DAY) {
        return ProjectionUtil.fixInclusiveTimeProjection(projected);
      }

      return projected;

    } else if (pred.isSetPredicate() && pred.op() == Expression.Operation.IN) {
      UnboundPredicate<Integer> projected =
          ProjectionUtil.transformSet(fieldName, pred.asSetPredicate(), apply);
      if (this != DAY) {
        return ProjectionUtil.fixInclusiveTimeProjection(projected);
      }
      return projected;

    } else if (pred.op() == Expression.Operation.RANGE_IN) {
      UnboundPredicate<Integer> projected;
      if (this != DAY) {
        projected =
            ((UnBoundCreator) pred).createTransformAppliedUnboundPred(apply, fieldName, true);
      } else {
        projected = ((UnBoundCreator) pred).createTransformAppliedUnboundPred(apply, fieldName);
      }
      return projected;
    }

    return null;
  }

  @Override
  public UnboundPredicate<Integer> projectStrict(String fieldName, BoundPredicate<Integer> pred) {
    if (pred.term() instanceof BoundTransform) {
      return ProjectionUtil.projectTransformPredicate(this, fieldName, pred);
    }

    if (pred.isUnaryPredicate()) {
      return Expressions.predicate(pred.op(), fieldName);

    } else if (pred.isLiteralPredicate()) {
      UnboundPredicate<Integer> projected =
          ProjectionUtil.truncateIntegerStrict(fieldName, pred.asLiteralPredicate(), apply);
      if (this != DAY) {
        return ProjectionUtil.fixStrictTimeProjection(projected);
      }

      return projected;

    } else if (pred.isSetPredicate() && pred.op() == Expression.Operation.NOT_IN) {
      UnboundPredicate<Integer> projected =
          ProjectionUtil.transformSet(fieldName, pred.asSetPredicate(), apply);
      if (this != DAY) {
        return ProjectionUtil.fixStrictTimeProjection(projected);
      }

      return projected;
    }

    return null;
  }

  @Override
  public String toHumanString(Type outputType, Integer value) {
    if (value == null) {
      return "null";
    }

    switch (granularity) {
      case YEARS:
        return TransformUtil.humanYear(value);
      case MONTHS:
        return TransformUtil.humanMonth(value);
      case DAYS:
        return TransformUtil.humanDay(value);
      default:
        throw new UnsupportedOperationException("Unsupported time unit: " + granularity);
    }
  }

  @Override
  public String toString() {
    return name;
  }

  @Override
  public String dedupName() {
    return "time";
  }
}

File: python/iceberg/api/transforms/dates.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import datetime

from .projection_util import (fix_inclusive_time_projection,
                              fix_strict_time_projection,
                              project_transform_predicate,
                              truncate_integer,
                              truncate_integer_strict)
from .transform import Transform
from .transform_util import TransformUtil
from ..expressions import (Expressions, Operation)
from ..expressions.predicate import BoundPredicate
from ..types.type import (Type, TypeID)
from ..types.types import (DateType, IntegerType)


class Dates(Transform):
    YEAR = "year"
    MONTH = "month"
    DAY = "day"

    EPOCH = datetime.datetime.utcfromtimestamp(0)
    SECONDS_IN_DAY = 86400

    HUMAN_FUNCS = {"year": lambda x: TransformUtil.human_year(x),
                   "month": lambda x: TransformUtil.human_month(x),
                   "day": lambda x: TransformUtil.human_day(x)}

    def __init__(self, granularity, name: str):
        super().__init__()
        if granularity not in (Dates.YEAR, Dates.MONTH, Dates.DAY):
            raise RuntimeError("Invalid Granularity: %s" % granularity)
        self._granularity = granularity
        self._name = name

    def apply(self, days):
        if days is None:
            return None

        if self._granularity == Dates.DAY:
            return days

        apply_func = getattr(TransformUtil, "diff_{}".format(self._granularity))
        if days >= 0:
            return apply_func(datetime.datetime.utcfromtimestamp(days * Dates.SECONDS_IN_DAY), Dates.EPOCH)
        else:
            return apply_func(datetime.datetime.utcfromtimestamp((days + 1) * Dates.SECONDS_IN_DAY), Dates.EPOCH) - 1

    def can_transform(self, another: Type):
        return another.type_id == TypeID.DATE

    def get_result_type(self, source_type):
        if self._granularity == Dates.DAY:
            return DateType.get()

        return IntegerType.get()

    # todo support preservesOrder and satisfiesOrderOf

    def project(self, name, predicate: BoundPredicate):
        from ..expressions.transform import BoundTransform
        if isinstance(predicate.term, BoundTransform):
            return project_transform_predicate(self, name, predicate)

        if predicate.is_unary_predicate:
            return Expressions.predicate(predicate.op, name)
        elif predicate.is_literal_predicate:
            projected = truncate_integer(name, predicate, self)
            if self._granularity != Dates.DAY:
                return fix_inclusive_time_projection(projected)
            return projected
        elif predicate.is_set_predicate and predicate.op == Operation.IN:
            raise NotImplementedError

        return None

    def project_strict(self, name, predicate):
        from ..expressions.transform import BoundTransform
        if isinstance(predicate.term, BoundTransform):
            return project_transform_predicate(self, name, predicate)

        if predicate.is_unary_predicate:
            return Expressions.predicate(predicate.op, name)
        elif predicate.is_literal_predicate:
            projected = truncate_integer_strict(name, predicate, self)
            if self._granularity != Dates.DAY:
                return fix_strict_time_projection(projected)
            return projected
        elif predicate.is_set_predicate and predicate.op == Operation.NOT_IN:
            raise NotImplementedError

        return None

    def to_human_string(self, value):
        if value is None:
            return "null"

        apply_func = Dates.HUMAN_FUNCS.get(self._granularity)
        if apply_func is not None:
            return apply_func(value)
        else:
            raise ValueError(f"Unsupported time unit: {self._granularity}")

    def __str__(self):
        return self._name

    def dedup_name(self):
        return "time"

    def __eq__(self, other):
        if id(self) == id(other):
            return True
        if other is None or not isinstance(other, Dates):
            return False

        return self._granularity == other._granularity and self._name == other._name
